File: src/benchmark/bench_full.nim
import ../oche
import std/[math, sequtils]

type 
  Body {.porche.} = object
    x, y, z, vx, vy, vz, mass: float64

  Vec3 {.porche.} = object
    x, y, z: float64

# [1] N-Body
proc nbodyNim(n: int): float64 {.porche.} =
  var bodies = @[
    Body(mass: 39.478), 
    Body(x: 4.84, y: -1.16, z: -0.103, vx: 0.606, vy: 2.812, vz: -0.025, mass: 0.037)
  ]
  let dt = 0.01
  for i in 0..<n:
    let dx = bodies[0].x - bodies[1].x
    let dy = bodies[0].y - bodies[1].y
    let dz = bodies[0].z - bodies[1].z
    let d2 = dx*dx + dy*dy + dz*dz
    let mag = dt / (d2 * sqrt(d2))
    bodies[0].vx -= dx * bodies[1].mass * mag
    bodies[0].x += dt * bodies[0].vx
  return bodies[0].x

# [2] Spectral Norm (แก้จุด error 'j' ตรงนี้ครับ)
proc A(i, j: int): float64 = 1.0 / (((i + j) * (i + j + 1) div 2 + i + 1).float64)

proc spectralNormNim(n: int): float64 {.porche.} =
  var u = newSeqWith(n, 1.0)
  var v = newSeq[float64](n)
  for _ in 0..9:
    for i in 0..<n:
      v[i] = 0.0
      for j in 0..<n: v[i] += A(i, j) * u[j]
    for i in 0..<n:
      u[i] = 0.0
      # แก้จาก 0..<j เป็น 0..<n
      for j in 0..<n: u[i] += A(j, i) * v[j]
  return sqrt(u[0])

# [3] Zero-Copy View
proc generateLargeArrayView(n: int): OcheBuffer[Vec3] {.porche.} =
  result = newOche[Vec3](n) 
  for i in 0..<n:
    result[i] = Vec3(x: i.float, y: i.float * 2.0, z: i.float * 3.0)

generatePython("nlib_full.py")

File: src/benchmark/bench_full.py
import time
import math
import nlib_full

def nbodyPython(n):
    x, vx, mass, dt = 4.84, 0.606, 0.037, 0.01
    for _ in range(n):
        d2 = x*x
        mag = dt / (d2 * math.sqrt(d2))
        vx -= x * mass * mag
        x += dt * vx
    return x

def spectralNormPython(n):
    u = [1.0] * n
    v = [0.0] * n
    for _ in range(10):
        for i in range(n):
            v[i] = sum(u[j] / ((i+j)*(i+j+1)//2 + i+1) for j in range(n))
        for i in range(n):
            u[i] = sum(v[j] / ((j+i)*(j+i+1)//2 + j+1) for j in range(n))
    return math.sqrt(u[0])

if __name__ == "__main__":
    print("==========================================")
    print("    🏎️  PORCHE FULL PERFORMANCE SUITE    ")
    print("==========================================\n")

    # Bench 1 & 2
    for name, py_fn, nim_fn, arg in [
        ("N-Body (1M)", nbodyPython, nlib_full.porche.nbodyNim, 1_000_000),
        ("Spectral Norm (500)", spectralNormPython, nlib_full.porche.spectralNormNim, 500)
    ]:
        print(f"[*] {name}")
        s = time.perf_counter()
        py_fn(arg)
        t_py = time.perf_counter() - s
        s = time.perf_counter()
        nim_fn(arg)
        t_nim = time.perf_counter() - s
        print(f"  - Python: {t_py:.4f}s | Nim: {t_nim:.4f}s | Speedup: {t_py/t_nim:.2f}x")

    # Bench 3: Zero-Copy
    SIZE = 10_000_000
    print(f"\n[*] Large Data Test ({SIZE:,} items)")
    s = time.perf_counter()
    view = nlib_full.porche.generateLargeArrayView(SIZE)
    print(f"  - Creation: {time.perf_counter() - s:.6f}s")

    try:
        import numpy as np
        s = time.perf_counter()
        np_arr = view.to_numpy() 
        if np_arr is not None:
            print(f"  - NumPy Zero-Copy: {time.perf_counter() - s:.6f}s")
            print(f"  - Shape: {np_arr.shape} | Sample: {np_arr[-1]}")
        else:
            print("  - to_numpy() returned None. Check nlib_full.py for 'return' statement.")
    except Exception as e:
        print(f"  - NumPy Error: {e}")